### ISSUE TITLE
RHEL8 Cannot perform replacement raidz1 vdev disk after device identification confused.

### ISSUE THREAD

**Issue by number201724** [openzfs/zfs#12631]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| RHEL
Distribution Version	| 8
Kernel Version	| RHEL 4.18.0-305.19.1.el8_4.x86_64
Architecture	|  x86_64
OpenZFS Version	| RHEL-based distro 2.0.6-1

### Describe the problem you're observing

I have ECC memory, but after the order of the disks is wrong, the status becomes UNAVAIL and cannot be replaced. 

### Describe how to reproduce the problem

```
  pool: data
 state: DEGRADED
status: One or more devices could not be used because the label is missing or
	invalid.  Sufficient replicas exist for the pool to continue
	functioning in a degraded state.
action: Replace the device using 'zpool replace'.
   see: https://openzfs.github.io/openzfs-docs/msg/ZFS-8000-4J
config:

	NAME                     STATE     READ WRITE CKSUM
	data                     DEGRADED     0     0     0
	  raidz1-0               DEGRADED     0     0     0
	    5469632888691130968  UNAVAIL      0     0     0  was /dev/sda1
	    sdb                  ONLINE       0     0     0
	    sdd                  ONLINE       0     0     0
	cache
	  sde                    FAULTED      0     0     0  corrupted data
	  sdf                    UNAVAIL      0     0     0

errors: No known data errors
```

1. zpool detach data sde sdf
```
  pool: data
 state: DEGRADED
status: One or more devices could not be used because the label is missing or
	invalid.  Sufficient replicas exist for the pool to continue
	functioning in a degraded state.
action: Replace the device using 'zpool replace'.
   see: https://openzfs.github.io/openzfs-docs/msg/ZFS-8000-4J
config:

	NAME                     STATE     READ WRITE CKSUM
	data                     DEGRADED     0     0     0
	  raidz1-0               DEGRADED     0     0     0
	    5469632888691130968  UNAVAIL      0     0     0  was /dev/sda1
	    sdb                  ONLINE       0     0     0
	    sdd                  ONLINE       0     0     0

errors: No known data errors
```
2. zpool offline data 5469632888691130968
```
  pool: data
 state: DEGRADED
status: One or more devices has been taken offline by the administrator.
	Sufficient replicas exist for the pool to continue functioning in a
	degraded state.
action: Online the device using 'zpool online' or replace the device with
	'zpool replace'.
config:

	NAME                     STATE     READ WRITE CKSUM
	data                     DEGRADED     0     0     0
	  raidz1-0               DEGRADED     0     0     0
	    5469632888691130968  OFFLINE      0     0     0  was /dev/sda1
	    sdb                  ONLINE       0     0     0
	    sdd                  ONLINE       0     0     0

errors: No known data errors
```
3. zpool replace data 5469632888691130968 sdc
```
invalid vdev specification
use '-f' to override the following errors:
/dev/sdc1 is part of active pool 'data'
```
4. fdisk /dev/sdc and g and w
```
NAME          MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
sda             8:0    0 931.5G  0 disk 
|-sda1          8:1    0 931.5G  0 part 
`-sda9          8:9    0     8M  0 part 
sdb             8:16   0   9.1T  0 disk 
|-sdb1          8:17   0   9.1T  0 part 
`-sdb9          8:25   0     8M  0 part 
sdc             8:32   0   9.1T  0 disk 
sdd             8:48   0   9.1T  0 disk 
|-sdd1          8:49   0   9.1T  0 part 
`-sdd9          8:57   0     8M  0 part 
sde             8:64   0 931.5G  0 disk 
|-sde1          8:65   0 931.5G  0 part 
`-sde9          8:73   0     8M  0 part 
sdf             8:80   0   1.1T  0 disk 
|-sdf1          8:81   0   600M  0 part /boot/efi
|-sdf2          8:82   0     1G  0 part /boot
`-sdf3          8:83   0   1.1T  0 part 
  |-rhel-root 253:0    0   1.1T  0 lvm  /
  `-rhel-swap 253:1    0   7.6G  0 lvm  [SWAP]
sr0            11:0    1  1024M  0 rom  

```
5. zpool replace data 5469632888691130968 sdc
`cannot replace 5469632888691130968 with sdc: sdc is busy, or device removal is in progress`
and lsblk
```
sda             8:0    0 931.5G  0 disk 
|-sda1          8:1    0 931.5G  0 part 
`-sda9          8:9    0     8M  0 part 
sdb             8:16   0   9.1T  0 disk 
|-sdb1          8:17   0   9.1T  0 part 
`-sdb9          8:25   0     8M  0 part 
sdc             8:32   0   9.1T  0 disk 
|-sdc1          8:33   0   9.1T  0 part 
`-sdc9          8:41   0     8M  0 part 
sdd             8:48   0   9.1T  0 disk 
|-sdd1          8:49   0   9.1T  0 part 
`-sdd9          8:57   0     8M  0 part 
sde             8:64   0 931.5G  0 disk 
|-sde1          8:65   0 931.5G  0 part 
`-sde9          8:73   0     8M  0 part 
sdf             8:80   0   1.1T  0 disk 
|-sdf1          8:81   0   600M  0 part /boot/efi
|-sdf2          8:82   0     1G  0 part /boot
`-sdf3          8:83   0   1.1T  0 part 
  |-rhel-root 253:0    0   1.1T  0 lvm  /
  `-rhel-swap 253:1    0   7.6G  0 lvm  [SWAP]
sr0            11:0    1  1024M  0 rom  

```
7. fdisk /dev/sdc and g and w
8. zpool replace -f data 5469632888691130968 sdc
```
zpool replace -f data 5469632888691130968 sdc
invalid vdev specification
the following errors must be manually repaired:
/dev/sdc1 is part of active pool 'data
```

```
sda             8:0    0 931.5G  0 disk 
|-sda1          8:1    0 931.5G  0 part 
`-sda9          8:9    0     8M  0 part 
sdb             8:16   0   9.1T  0 disk 
|-sdb1          8:17   0   9.1T  0 part 
`-sdb9          8:25   0     8M  0 part 
sdc             8:32   0   9.1T  0 disk 
|-sdc1          8:33   0   9.1T  0 part 
`-sdc9          8:41   0     8M  0 part 
sdd             8:48   0   9.1T  0 disk 
|-sdd1          8:49   0   9.1T  0 part 
`-sdd9          8:57   0     8M  0 part 
sde             8:64   0 931.5G  0 disk 
|-sde1          8:65   0 931.5G  0 part 
`-sde9          8:73   0     8M  0 part 
sdf             8:80   0   1.1T  0 disk 
|-sdf1          8:81   0   600M  0 part /boot/efi
|-sdf2          8:82   0     1G  0 part /boot
`-sdf3          8:83   0   1.1T  0 part 
  |-rhel-root 253:0    0   1.1T  0 lvm  /
  `-rhel-swap 253:1    0   7.6G  0 lvm  [SWAP]
sr0            11:0    1  1024M  0 rom  

```

**Comment by rincebrain** [comment]
Your title seems to have some things that aren't really related to the issue at hand.

It seems like you need a wipefs on the disk before it will allow you to replace with it, even with -f, so that might be a bug? Kind of messy because AIUI devices like cache devices have less labeling on them to figure out they're in use or not, and most people don't have a use case where they would swap from using something as L2ARC to a data disk.

What does zpool status say after initially repartitioning on sdc? Does wipefs help you?

**Comment by number201724** [comment]
The first is that the device ID is out of order, and then I need to reconfigure. 
`5469632888691130968  UNAVAIL      0     0     0  was /dev/sda1`

The correct location of this disk should be `sdc`

zfs prompts that I need to use zfs replace to replace this device.

`zfs replace data 5469632888691130968 sdc`

Then the replacement failed. After I repartitioned, the replacement still failed again.


**What does zpool status say after initially repartitioning on sdc?** 
```
invalid vdev specification
use '-f' to override the following errors:
/dev/sdc1 is part of active pool 'data'
```

**Does wipefs help you?**
I don't use wipefs, I just need to reconfigure the correct device id and let him resynchronize. I deleted the cache disk and will append it after the configuration is correct.
